tests/Crypto/cryptotest.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2017 Jolla Ltd.
+ * Contact: Bea Lam <bea.lam@jollamobile.com>
+ * All rights reserved.
+ * BSD 3-Clause License, see LICENSE.
+ */
+
+#include "cryptotest.h"
+
+#include "Crypto/plugininforequest.h"
+
+#include <QFile>
+#include <QtDebug>
+
+using namespace Sailfish::Crypto;
+
+namespace {
+
+QString findPluginName(const char *name, const QString &defaultValue)
+{
+    QByteArray envVarValue = qgetenv(name);
+    QString value = envVarValue.isEmpty() ? defaultValue : QString::fromLatin1(envVarValue);
+    qDebug() << "\t" << name << "=" << value;
+    return value;
+}
+
+}
+
+CryptoTest::CryptoTest(QObject *parent)
+    : QObject(parent)
+{
+    qDebug() << "Test will run with these plugins:";
+    m_testPluginCrypto = findPluginName("TEST_PLUGIN_CRYPTO", CryptoManager::DefaultCryptoPluginName + QLatin1String(".test"));
+    m_testPluginCryptoStorage = findPluginName("TEST_PLUGIN_CRYPTO_STORAGE", CryptoManager::DefaultCryptoStoragePluginName + QLatin1String(".test"));
+    m_testPluginStorage = findPluginName("TEST_PLUGIN_STORAGE", Sailfish::Secrets::SecretManager::DefaultStoragePluginName + QLatin1String(".test"));
+    m_testPluginEncryptedStorage = findPluginName("TEST_PLUGIN_ENCRYPTED_STORAGE", Sailfish::Secrets::SecretManager::DefaultEncryptedStoragePluginName + QLatin1String(".test"));
+    m_testPluginAuth = findPluginName("TEST_PLUGIN_AUTH", Sailfish::Secrets::SecretManager::DefaultAuthenticationPluginName + QLatin1String(".test"));
+    m_testPluginInAppAuth = findPluginName("TEST_PLUGIN_IN_APP_AUTH", Sailfish::Secrets::SecretManager::InAppAuthenticationPluginName + QLatin1String(".test"));
+    m_testPluginEncryption = findPluginName("TEST_PLUGIN_ENCRYPTION", Sailfish::Secrets::SecretManager::DefaultEncryptionPluginName + QLatin1String(".test"));
+    m_testPluginUsbToken = findPluginName("TEST_PLUGIN_USB_TOKEN", QStringLiteral("org.sailfishos.secrets.plugin.cryptostorage.exampleusbtoken.test"));
```

Comment:
While I like this idea in theory, in practice different plugins will have different levels of support for different operations and permutations of arguments - which means that specifying some different plugin via an environment variable is likely to result in a variety of unit test failures which shouldn't actually be considered failures (because different plugins support different semantics).  An alternative might be to pass the plugin names as QTest columns in the various _data() methods, along with expected results for different operations.  That way, other plugin implementors can add their own unit tests to enforce appropriate behaviour for their plugin simply by adding a data row to the appropriate _data() column.

Reply:
> specifying some different plugin via an environment variable is likely to result in a variety of 
> unit test failures which shouldn't actually be considered failures (because different plugins 
> support different semantics)

What sort of semantics? Does that mean the unit test itself may have to be changed per-plugin? How should it be reconfigured or skipped in those cases?

I did want to avoid adding _data() per-plugin as it means a lot of duplicated setup code, since QtTest doesn't support adding a set of test data on top of all current sets of test data (i.e. adding a set of providers across all the current test data configurations) but we can do that if necessary. 

If multiple plugins need to be supported across every unit test, then I suggest having an ini file that specifies the plugins for each plugin type (TEST_PLUGIN_AUTH etc).

Reply:
Yes, the unit tests need to test different things on a per-plugin basis, most likely.  That is, some plugins won't support some operations, or some block modes, or whatever, while others will.

So we cannot simply use an environment variable to specify the plugin name, in most cases, as each plugin may implement different things.  E.g. what if the OMP plugin doesn't support AES at all, but only supports GOST?  Or doesn't support SHA256 but supports SHA3_512 only?  etc.  Thus, the plugin name needs to be specified in the test data row, along with every input datum (e.g. algorithm, block mode, digest function, data to encrypt, etc) and every output datum (expected values, etc) .

What do you mean by "adding _data() per-plugin" there?  I just meant something like: each test should test some specific functionality.  So each data row for that test should include:

a) the name of the plugin to be tested
c) various inputs to specify the operation parameters
d) the expected result values for the given inputs

If a single unit test tests multiple independent features (e.g. the stored key requests which test generating stored keys, filtering stored keys, and using stored keys via encrypt/decrypt) it may be that a plugin supports some of those stages (e.g. generating and storing, and using the stored keys via encrypt/decrypt) but not others (e.g. it may not support filtering stored keys), and in that case the "does the plugin support this feature" flag needs to be passed as a row data also.  For example:

a) the name of the plugin to be tested
b) input data required to test generating stored keys
c) the expected result from that operation
d) whether the plugin supports filtering
e) input data required to test filtering
f) the expected result from filtering
g) whether the plugin supports encrypt operations
h) input data required to test encrypt operations
i) the expected result from encrypting
j) whether the plugin supports decrypt operations
k) input data required to test decrypt operations
l) the expected result from decrypting.

Then the test itself should look something like:

```
generateStoredKey(input params);
QCOMPARE(result, expectedGSKResult);

if (pluginSupportsFiltering) {
    filterKeys(someFilterParameters);
    QCOMPARE(results, expectedFilterResults);
}

if (pluginSupportsEncrypt) {
    encrypt(storedKey, inputData);
    QCOMPARE(results, expectedEncryptResults);

    if (pluginSupportsDecrypt) {
        decrypt(storedKey, encryptedData);
        QCOMPARE(decryptedData, originalEncryptInputData);
    }
}
```

Reply:
It previously existed, Joona asked me to remove it, as it complicated the API, and clients should know the semantics of the plugin they are attempting to interact with.  I actually agree with his reasoning about that: because the semantics of different plugins are different, merely reporting it programmatically is unlikely to be enough to allow a client to write code which has no knowledge of that plugin anyway (a good example is: what if the plugin requires a specific customParameter to be set?  do we also need some way to programmatically report that?)

Also, previously we reported it as a map from algorithm (e.g. AES) to QVector<BlockMode> or to QVector<SomeEncryptionPadding> etc - but this doesn't tell the complete picture, as a plugin may support AES with block mode CBC but only with some specific padding (like None), and also AES with a different block mode GCM, but only with some other specific padding (like PKCS#11 or whatever).  Basically, in order to report it effectively, programmatically, we would have some insanely complex lookup table (which then has to be defined by the plugin implementor!) in order to be useful...

Reply:
Yes it sounds like it would be quite complex to have a plugin API that provides useful information about exactly what the plugin supports, so it would be too much trouble.

> etc - but this doesn't tell the complete picture, as a plugin may support AES with block mode 
> CBC but only with some specific padding (like None), and also AES with a different block mode 
> GCM, but only with some other specific padding (like PKCS#11 or whatever). 

If the configuration is purely just pluginSupportsEncrypt, pluginSupportsDecrypt etc. then we wouldn't be able to support these sorts of requirements in the tests. Is the plan to leave out this sort of detailed configuration in the tests for now?

Reply:
It's not needed, because that configuration is specified in the test itself.  That is, if you're writing some plugin and your plugin doesn't support AES CBC, then you either don't add a test data row for AES CBC, or alternatively you add a test data row for AES CBC which ensures that the return value is a failure with the expected error code.